README
Name
    nginx-xsltproc-module - an extended version of the standard module
    HttpXsltModule that allows you to specify a stylesheet in fastcgi
    script by passing a specified header.

Synopsis
    #nginx.conf
        location / {
            xsltproc        on;

            xsltproc_stylesheet_caching on;
            xsltproc_stylesheet_root /var/www/example.com/template;
            xsltproc_stylesheet_check_if_modify on;

            xsltproc_profiler on;
            xsltproc_profiler_stylesheet /var/www/example.com/template/profiler.xslt;
            xsltproc_profiler_repeat on;

            xsltproc_memcached on;
            xsltproc_memcached_server 1.1.1.1:11211;
            xsltproc_memcached_server 2.2.2.2:11211;
            xsltproc_memcached_key_prefix site1_;
            xsltproc_memcached_key_auto on;
            xsltproc_memcached_expire 60m;

            xsltproc_types  application/xml;

            xml_entities    /site/dtd/entities.dtd;

            fastcgi_pass    127.0.0.1:9009;
            fastcgi_index   index.fcgi;
            fastcgi_param   SCRIPT_FILENAME  /scripts$fastcgi_script_name;
            include         fastcgi_params;
        }

    #index.fcgi
        #!/usr/bin/perl

        use strict;
        use warnings;

        use CGI::Fast;

        my $socket  = FCGI::OpenSocket("127.0.0.1:9009", 256);
        my $request = FCGI::Request(\*STDIN, \*STDOUT, \*STDERR, \%ENV, $socket);

        while ($request->Accept() >= 0) {

            print
                "X-Xslt-Stylesheet: /xslt/test.xslt?param1=1&param2='test'\r\n",
                "Content-type: application/xml\r\n\r\n",
                "<root></root>";
        }

Description
    This module extends the standard module HttpXsltModule to specify
    a stylesheet in fastcgi script by passing a specified header.

    Header must be specified as:
    X-Xslt-Stylesheet: <stylesheet_path>[?<param>=<value>[&<param>=<value>...]]

    Stylesheet path is root-relative.

    Parameter values must be escaped with the URI escaping mechanism.

Directives
    xsltproc
        syntax: *xsltproc on|off*

        default: *xsltproc off*

        context: *http, server, location, if in location*

        Turning this directive on will enable XSLT processing.

    xsltproc_stylesheet_root
        syntax: *xsltproc_stylesheet_root <root_path>*

        default: *no*

        context: *http, server, location, if in location*

        Sets the root directory for xslt-templates.

    xsltproc_stylesheet_caching
        syntax: *xsltproc_stylesheet_caching on|off*

        default: *xsltproc_stylesheet_caching off*

        context: *http, server, location, if in location*

        Turning this directive on will enable stylesheet caching.

    xsltproc_stylesheet_check_if_modify
        syntax: *xsltproc_stylesheet_check_if_modify on|off*

        default: *xsltproc_stylesheet_check_if_modify off*

        context: *http, server, location, if in location*

        Turning this directive on will check if a stylesheet has been modified
        on every request and will update cache if need.

    xsltproc_profiler
        syntax: *xsltproc_profiler on|off*

        default: *xsltproc_profiler off*

        context: *http, server, location, if in location*

        Turning this directive on will enable collection the profile information.

    xsltproc_profiler_stylesheet
        syntax: *xsltproc_profiler_stylesheet <stylesheet_path>*

        default: *no*

        context: *http, server, location, if in location*

        If parameter is specified, the profile information added with this template
        in the resulting HTML document.

        The structure of profile information:
            <profiler parse_header_time="44" parse_body_time="10">
                <stylesheet uri="main.xsl" time="111">
                  <profile>
                    <template rank="1" name="" match="*" mode="" calls="20" time="444" />
                    ...
                  </profile>
                  <document>
                    <root>
                      ...
                    </root>
                  </document>
                  <params>
                    <param name="name1" value="'value1'" />
                    ...
                  </params>
                </stylesheet>
                ...
            </profiler>

        An example the result document:
            <html>
                <head>
                    ...
                </head>
                <body>
                    ...
                    <!-- the result of applying the profiler stylesheet -->
                    <div id="profiler" class="profiler">
                        ...
                    </div>
                </body>
            </html

    xsltproc_profiler_repeat
        syntax: *xsltproc_profiler_repeat on|off*

        default: *xsltproc_profiler_repeat off*

        context: *http, server, location, if in location*

        Turning this directive on will run the transformation 20 times.

    xsltproc_memcached
        syntax: *xsltproc_memcached on|off*

        default: *xsltproc_memcached off*

        context: *http, server, location, if in location*

        Turning this directive on will enable using memcahed.

    xsltproc_memcached_server <name:[port]>
        syntax: *xsltproc_memcached_server <url>*

        default: *none*

        context: *http, server, location, if in location*

        Push a server into the list of memcahed servers.

    xsltproc_memcached_key_prefix <prefix>
        syntax: *xsltproc_memcached_key_prefix <prefix>*

        default: *none*

        context: *http, server, location, if in location*

        The memcached key is "$prefix..."

    xsltproc_memcached_key_auto
        syntax: *xsltproc_memcached_key_auto on|off*

        default: *xsltproc_memcached_key_auto off*

        context: *http, server, location, if in location*

        Turning this directive on will calculate a memcached key as md5sum
        of the body + stylesheet info (uri, mtime, params).

    xsltproc_memcached_expire <time>
        syntax: *xsltproc_memcached_expire <time>*

        default: *0*

        context: *http, server, location, if in location*

        Expiration time.

    xml_entities
        syntax: *xml_entities <path>*

        default: *no*

        context: *http, server, location, if in location*

        Specifies the DTD file which describes symbolic elements (xml entities).
        This file is compiled at the stage of configuration. For technical
        reasons it's not possible to specify entities in the XML being processed,
        therefore they are ignored, but this specially assigned file is used
        instead. In this file it is not necessary to describe structure of
        processed XML, it is sufficient only to declare necessary symbolic
        elements, for example:
            <! ENTITY of nbsp "  ">

    xsltproc_types
        syntax: *xsltproc_types mime-type [mime-type...]*

        default: *xsltproc_types text/xml*

        context: *http, server, location, if in location*

        Permit processing responses with specified MIME-types in addition to "text/xml".
        If XSLT output mode is HTML, then the response MIME-type changes to "text/HTML".

Using XSLT functions
    <xsl:stylesheet version="1.0"
        xmlns:ngx="http://nginx.org/xsltproc"
        extension-element-prefixes="ngx"
    >

    ...

    </xsl:stylesheet>

    ngx:join
        syntax: *string ngx:join(node-set, sep)*

        Returns a string created by concatenating the string arguments and using
        the sep argument as the separator.

        example: ngx:join('string1', 'string2', /root/item, ', ')

        xml: '<root><item>item1</item><item>item2</item></root>'

        result: 'string1, string2, item1, item2'

    ngx:uc
        syntax: *string ngx:uc(string)*

        Converts the string argument to upper-case

        example: ngx:uc('abcDEF')

        result: 'ABCDEF'

    ngx:lc
        syntax: *string ngx:lc(string)*

        Converts the string argument to lower-case

        example: *ngx:lc('abcDEF')*

        result: 'abcdef'

    ngx:ltrim
        syntax: *string ngx:ltrim(string)*

        Strip whitespace from the beginning of a string

        example: ngx:ltrim('   string')

        result: 'string'

    ngx:rtrim
        syntax: *string ngx:rtrim(string)*

        Strip whitespace from the end of a string

        example: ngx:rtrim('string    ')

        result: 'string'

    ngx:trim
        syntax: *string ngx:trim(string)*

        Strip whitespace from the beginning and end of a string

        example: ngx:trim('    string    ')

        result: 'string'


Installation
    Grab the nginx source code from nginx.org (<http://nginx.org/>), for
    example, the version 1.0.11, and then build the source with this module:

        wget 'http://nginx.org/download/nginx-1.0.11.tar.gz'
        tar -xzvf nginx-1.0.11.tar.gz
        cd nginx-1.0.11/

        ./configure --add-module=/path/to/nginx-xsltproc-module

        make -j2
        make install

    Download the latest version of the release tarball of this module from
    nginx-xsltproc-module file list
    (<https://github.com/yoreek/nginx-xsltproc-module/tags>).

Copyright & License
    This module is based on the HttpXsltModule module in the Nginx 1.0.11 sources.
    The original code is copyright Igor Sysoev.

    Copyright (c) 2012, Yuriy Ustyushenko <yoreel@yahoo.com>.

    This module is licensed under the terms of the BSD license.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

    *   Redistributions of source code must retain the above copyright
        notice, this list of conditions and the following disclaimer.

    *   Redistributions in binary form must reproduce the above copyright
        notice, this list of conditions and the following disclaimer in the
        documentation and/or other materials provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
    IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
    TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
    TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


CHANGES
2021-01-11    VERSION 0.17
	* fix compilation bug with undeclared FALSE macro

2014-02-17    VERSION 0.15
    * libmemcached is no longer required

2013-04-08    VERSION 0.14
    * fixed memory free bug

2013-03-01    VERSION 0.13
    * fixed bug with "xsltproc_profiler_repeat" option

2013-02-28    VERSION 0.12
    * fixed compiler warnings

2012-04-01    VERSION 0.10
    * new feature:
      - new option 'xsltproc_profiler_repeat' for the profiler.
      - caching results of XSLT transformation with the libMemcached library.

2012-03-19    VERSION 0.9
    * fix bug when compiling with ICU lib.

2012-03-17    VERSION 0.8
    * added some xslt functions (uc, lc, join, ltrim, rtrim, trim).

2012-03-10    VERSION 0.7
    * added "xslproc_stylesheet_root" option.

2012-03-01    VERSION 0.6
    * fix bug with merging configuration.

2012-02-27    VERSION 0.5
    * fix many bugs.
    * instead of "xsltproc_stylesheet_profiler" option made two options
    "xsltproc_profiler" and "xsltproc_profiler_stylesheet".

2012-02-26    VERSION 0.4
    * added performance tests.
    * fix bug when parameters for stylesheet is not defined.

2012-02-25    VERSION 0.3
    * added stylesheet profiler.
    * renamed header name "X-Xslt" to "X-Xslt-Stylesheet".

2012-02-06    VERSION 0.2
    * added documents and keys caching.

2012-01-02    VERSION 0.1
    * Initial release.


src/ngx_http_xsltproc_xslt.c
#include "ngx_http_xsltproc_core.h"
#include "ngx_http_xsltproc_xslt.h"

/*static xsltTransformContextPtr tr_ctxt = NULL;*/
static ngx_log_t *xslt_log = NULL;

#if (NGX_HTTP_XSLTPROC_PROFILER)
/*
 * <profiler parse_header_time="20" parse_body_time="44">
 *   <stylesheet uri="main.xsl" time="444">
 *     <profile>
 *       <template name="" match="*" mode="" ... />
 *       ...
 *     </profile>
 *     <document>
 *       <root>
 *         ...
 *       </root>
 *     </document>
 *     <params>
 *       <param name="name1" value="'value1'" />
 *       ...
 *     </params>
 *   </stylesheet>
 *   ...
 * </profiler>
 */

void ngx_http_xsltproc_xslt_profiler_init(ngx_http_xsltproc_profiler_t *profiler) {

    xmlNodePtr root;
    char       strbuf[100];

    profiler->parse_body_time = xsltTimestamp() - profiler->parse_body_start;

    profiler->summary_profile_info = xmlNewDoc((const xmlChar*) "1.0");
    profiler->summary_profile_info->encoding = (const xmlChar*) xmlStrdup((const xmlChar*) "utf-8");

    root = xmlNewDocNode(profiler->summary_profile_info, NULL, BAD_CAST "profiler", NULL);
    xmlDocSetRootElement(profiler->summary_profile_info, root);

    sprintf(strbuf, "%ld", profiler->parse_header_time);
    xmlSetProp(root, BAD_CAST "parse_header_time", BAD_CAST strbuf);

    sprintf(strbuf, "%ld", profiler->parse_body_time);
    xmlSetProp(root, BAD_CAST "parse_body_time", BAD_CAST strbuf);

    sprintf(strbuf, "%d", profiler->repeat);
    xmlSetProp(root, BAD_CAST "repeat", BAD_CAST strbuf);
}

void ngx_http_xsltproc_xslt_profiler_done(ngx_http_xsltproc_profiler_t *profiler,
                                          xsltStylesheetPtr stylesheet, xmlDocPtr doc) {

    xmlNodePtr root;
    xmlDocPtr  res;

    res = xsltApplyStylesheet(stylesheet, profiler->summary_profile_info, NULL);

    if (res) {
        root = xmlDocCopyNode(xmlDocGetRootElement(res), doc, 1);

        if (root) {
            xmlAddChild(xmlDocGetRootElement(doc)->last, root);
        }
        else {
            ngx_log_error(NGX_LOG_ERR, xslt_log, 0,
                           "ngx_http_xsltproc_xslt_profiler_done: no root in profile info");
        }

        xmlFreeDoc(res);
    }
    else {
        ngx_log_error(NGX_LOG_ERR, xslt_log, 0,
                       "ngx_http_xsltproc_xslt_profiler_done: no profile info");
    }

    xmlFreeDoc(profiler->summary_profile_info);
}

static void ngx_http_xsltproc_xslt_profiler_add_info(ngx_http_xsltproc_profiler_t *profiler,
    ngx_http_xsltproc_xslt_stylesheet_t *xslt_stylesheet, const char **params,
    long spent, xmlDocPtr doc, xmlDocPtr profile_info) {

    xmlNodePtr root, child, child2, child3;
    char       strbuf[100];

    root = xmlDocGetRootElement(profiler->summary_profile_info);

    /* add stylesheet info */
    child = xmlNewChild(root, NULL, BAD_CAST "stylesheet", NULL);
    xmlSetProp(child, BAD_CAST "uri", BAD_CAST xslt_stylesheet->uri);

    sprintf(strbuf, "%ld", spent);
    xmlSetProp(child, BAD_CAST "time", BAD_CAST strbuf);

    /* add profile info */
    xmlAddChild(child, xmlDocCopyNode(xmlDocGetRootElement(profile_info), profiler->summary_profile_info, 1));

    /* add document */
    child2 = xmlNewChild(child, NULL, BAD_CAST "document", NULL);
    xmlAddChild(child2, xmlDocCopyNode(xmlDocGetRootElement(doc), profiler->summary_profile_info, 1));

    /* add params */
    child2 = xmlNewChild(child, NULL, BAD_CAST "params", NULL);
    if (params != NULL) {
        for(; *params != '\0'; params++) {
            child3 = xmlNewChild(child2, NULL, BAD_CAST "param", NULL);

            xmlSetProp(child3, BAD_CAST "name", BAD_CAST *params);
            params++;
            xmlSetProp(child3, BAD_CAST "value", BAD_CAST *params);
        }
    }
}
#endif

#if (NGX_HTTP_XSLTPROC_XSLT_DOCUMENT_CACHING && NGX_HTTP_XSLTPROC_XSLT_KEYS_CACHING)
static void ngx_http_xsltproc_xslt_backup_keys(
    ngx_http_xsltproc_xslt_stylesheet_t *xslt_stylesheet, xsltDocumentPtr *doc_list)
{
    ngx_http_xsltproc_xml_document_extra_info_t *doc_extra_info;
    xmlDocPtr                                    doc;
    xsltDocumentPtr                              xslt_document;

#ifdef NGX_DEBUG
    ngx_log_error_core(NGX_LOG_DEBUG, xslt_log, 0,
                   "ngx_http_xsltproc_xslt_backup_keys: stylesheet %s",
                   xslt_stylesheet->uri);
#endif

    for ( ;; ) {
        xslt_document = *doc_list;

        if (xslt_document == NULL || xslt_document->main)
            break;

        doc            = xslt_document->doc;
        doc_extra_info = doc->_private;

#ifdef NGX_DEBUG
        ngx_log_error_core(NGX_LOG_DEBUG, xslt_log, 0,
                   "ngx_http_xsltproc_xslt_backup_keys: document %s",
                   doc->URL);
#endif

        ngx_http_xsltproc_xslt_keys_cache_put(xslt_stylesheet->uri, xslt_stylesheet->mtime,
            (char *) doc->URL, doc_extra_info->mtime, xslt_document);

        *doc_list = xslt_document->next;
    }
}

static void ngx_http_xsltproc_xslt_restore_keys(ngx_http_xsltproc_xslt_stylesheet_t *xslt_stylesheet,
    xsltDocumentPtr *doc_list)
{
    ngx_http_xsltproc_list_t      *el, xslt_keys_list;
    ngx_http_xsltproc_xslt_keys_t *xslt_keys;
    xsltDocumentPtr xslt_document;

#ifdef NGX_DEBUG
    ngx_log_error_core(NGX_LOG_DEBUG, xslt_log, 0,
                   "ngx_http_xsltproc_xslt_restore_keys: stylesheet %s",
                   xslt_stylesheet->uri);
#endif

    ngx_http_xsltproc_xslt_keys_cache_get(&xslt_keys_list, xslt_stylesheet->uri, xslt_stylesheet->mtime);

    for (
        el  = ngx_http_xsltproc_list_first(&xslt_keys_list);
        el != ngx_http_xsltproc_list_end(&xslt_keys_list);
        el  = ngx_http_xsltproc_list_next(el)
    ) {
        xslt_keys = (ngx_http_xsltproc_xslt_keys_t *) el;

#ifdef NGX_DEBUG
    ngx_log_error_core(NGX_LOG_DEBUG, xslt_log, 0,
                   "ngx_http_xsltproc_xslt_restore_keys: document %s",
                   xslt_keys->document_uri);
#endif

        if ((xslt_document = ngx_http_xsltproc_malloc(sizeof(xsltDocument))) == NULL) {
            return;
        }
        memcpy(xslt_document, xslt_keys->xslt_document, sizeof(xsltDocument));

        xslt_document->next = *doc_list;
        *doc_list           = xslt_document;
    }

    ngx_http_xsltproc_xslt_keys_list_free(&xslt_keys_list, XSLT_KEYS_LIST_FREE_NONE);
}
#endif

#if (NGX_HTTP_XSLTPROC_PROFILER)
static void ngx_http_xsltproc_xslt_reset_profile_info(xsltTransformContextPtr ctxt) {
    xsltTemplatePtr   template;
    xsltStylesheetPtr style;

    style = ctxt->style;

    while (style != NULL) {
        template = style->templates;

        while (template != NULL) {
            template->nbCalls = 0;
            template->time    = 0;

            template = template->next;
        }

        style = xsltNextImport(style);
    }
}
#endif

#if (NGX_HTTP_XSLTPROC_PROFILER)
xmlDocPtr ngx_http_xsltproc_xslt_transform(ngx_http_xsltproc_xslt_stylesheet_t *xslt_stylesheet,
    xmlDocPtr doc, const char **params, int profiler_enabled, ngx_http_xsltproc_profiler_t *profiler)
{
    xmlDocPtr               result = NULL;
    xsltTransformContextPtr ctxt;
    xmlDocPtr               profile_info;
    long                    start = 0, spent = 0;
    int                     i = 1, repeat = 1;

    if (profiler_enabled == 1 && profiler != NULL) {
        repeat = profiler->repeat;
    }

    for (i = 1; i <= repeat; i++) {
        ctxt = (xsltTransformContextPtr) xsltNewTransformContext(xslt_stylesheet->stylesheet, doc);

#if (NGX_HTTP_XSLTPROC_XSLT_DOCUMENT_CACHING && NGX_HTTP_XSLTPROC_XSLT_KEYS_CACHING)
        ngx_http_xsltproc_xslt_restore_keys(xslt_stylesheet, &ctxt->docList);
#endif

        if (profiler_enabled == 1) {
            /* reset internal template counters */
            if (i == 1) {
                ngx_http_xsltproc_xslt_reset_profile_info(ctxt);
                start = xsltTimestamp();
            }

            result = (xmlDocPtr) xsltApplyStylesheetUser(xslt_stylesheet->stylesheet, doc, params, NULL, stderr, ctxt);

            if (result == NULL)
                break;

            spent = xsltTimestamp() - start;

            if (profiler != NULL && i == repeat) {
                profile_info = xsltGetProfileInformation(ctxt);
                if (profile_info != NULL) {
                    ngx_http_xsltproc_xslt_profiler_add_info(profiler, xslt_stylesheet,
                                                             params, spent, doc, profile_info);
                    xmlFreeDoc(profile_info);
                }
                else {
                    xmlFreeDoc(result);
                    result = NULL;

                    ngx_log_error(NGX_LOG_ERR, xslt_log, 0,
                                  "ngx_http_xsltproc_xslt_transform: no profile info");
                }
            }
        }
        else {
            result = (xmlDocPtr) xsltApplyStylesheetUser(xslt_stylesheet->stylesheet, doc, params, NULL, NULL, ctxt);
        }

#if (NGX_HTTP_XSLTPROC_XSLT_DOCUMENT_CACHING && NGX_HTTP_XSLTPROC_XSLT_KEYS_CACHING)
        ngx_http_xsltproc_xslt_backup_keys(xslt_stylesheet, &ctxt->docList);
#endif

        xsltFreeTransformContext(ctxt);

        if (i < repeat && result != NULL) {
            xmlFreeDoc(result);
            result = NULL;
        }
    }

    return result;
}
#else
xmlDocPtr ngx_http_xsltproc_xslt_transform(ngx_http_xsltproc_xslt_stylesheet_t *xslt_stylesheet,
    xmlDocPtr doc, const char **params)
{
    xmlDocPtr               result;
    xsltTransformContextPtr ctxt;

    ctxt = (xsltTransformContextPtr) xsltNewTransformContext(xslt_stylesheet->stylesheet, doc);

#if (NGX_HTTP_XSLTPROC_XSLT_DOCUMENT_CACHING && NGX_HTTP_XSLTPROC_XSLT_KEYS_CACHING)
    ngx_http_xsltproc_xslt_restore_keys(xslt_stylesheet, &ctxt->docList);
#endif

    result = (xmlDocPtr) xsltApplyStylesheetUser(xslt_stylesheet->stylesheet, doc, params, NULL, NULL, ctxt);

#if (NGX_HTTP_XSLTPROC_XSLT_DOCUMENT_CACHING && NGX_HTTP_XSLTPROC_XSLT_KEYS_CACHING)
    ngx_http_xsltproc_xslt_backup_keys(xslt_stylesheet, &ctxt->docList);
#endif

    xsltFreeTransformContext(ctxt);

    return result;
}
#endif

void ngx_http_xsltproc_xslt_init(ngx_log_t *log) {
    xslt_log = log;

    /*xsltSetGenericDebugFunc(stderr, NULL);*/
    exsltRegisterAll();
    /*xsltRegisterTestModule();*/
    /*xsltDebugDumpExtensions(NULL);*/

    ngx_http_xsltproc_xslt_function_init(log);
    ngx_http_xsltproc_xslt_stylesheet_init(log);

#if (NGX_HTTP_XSLTPROC_XSLT_DOCUMENT_CACHING)
    ngx_http_xsltproc_xslt_document_init(log);
    ngx_http_xsltproc_xslt_keys_init(log);
#endif

}

void ngx_http_xsltproc_xslt_cleanup(void) {
#if (NGX_HTTP_XSLTPROC_XSLT_DOCUMENT_CACHING)
    ngx_http_xsltproc_xslt_keys_destroy();
    ngx_http_xsltproc_xslt_document_destroy();
#endif

    ngx_http_xsltproc_xslt_stylesheet_destroy();

    xsltCleanupGlobals();
}
